lxd/gendocs/fileserver.go

```diff
@@ -0,0 +1,211 @@
+var preferredEncodings = []string{"gzip", "identity"}
+
+// File extension to use for different encodings.
+func extensionForEncoding(encname string) string {
+	switch encname {
+	case "gzip":
+		return ".gz"
+	case "identity":
+		return ""
+	}
+
+	return ""
+}
+
+type fileHandler struct {
+	root FileSystem
+}
+
+// FileServer creates our custom http.FileServer
+// that can support the search for multiple compressed/uncompressed
+// versions of a file and select the best.
+func FileServer(root FileSystem) http.Handler {
+	return &fileHandler{root}
+}
+
+// openAndStat simply tries to open the file and stat it.
+func (f *fileHandler) openAndStat(path string) (http.File, os.FileInfo, error) {
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+
+	file, err := f.root.Open(path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	info, err := file.Stat()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return file, info, nil
+}
+
+const (
+	acceptEncodingHeader  = "Accept-Encoding"
+	contentEncodingHeader = "Content-Encoding"
+	contentLengthHeader   = "Content-Length"
+	rangeHeader           = "Range"
+	varyHeader            = "Vary"
+)
+
+// Find the best file to serve based on the client's Accept-Encoding, and which
+// files actually exist on the filesystem. If no file was found that can satisfy
+// the request, return an error.
+func (f *fileHandler) findBestFile(w http.ResponseWriter, r *http.Request, name string) (http.File, os.FileInfo, error) {
+	ae := r.Header.Get(acceptEncodingHeader)
+	if ae == "" {
+		// Try to read the file without any encoding.
+		return f.openAndStat(name)
+	}
+
+	// See what possible encodings we can send by looking for files
+	var available []string
+	var fnames []string
+	for _, posenc := range preferredEncodings {
+		ext := extensionForEncoding(posenc)
+		fname := name + ext
+		fnames = append(fnames, fname)
+		if f.root.Exists(fname) {
+			available = append(available, posenc)
+		}
+	}
+
+	if len(available) == 0 {
+		// If no asset (compressed or uncompressed) has been found,
+		// read the path as is and return. We don't want to return an error now
+		// if the path is a directory
+		return f.openAndStat(name)
+	}
+
+	ext := extensionForEncoding(available[0]) // choose the best encoding available.
+	file, info, err := f.openAndStat(name + ext)
+	if err != nil {
+		// At this point, we know the file exists
+		// and is not a directory, so if we can't open it, return an error
+		return nil, nil, err
+	}
+
+	wHeader := w.Header()
+	wHeader[contentEncodingHeader] = []string{available[0]}
+	wHeader.Add(varyHeader, acceptEncodingHeader)
+
+	if len(r.Header[rangeHeader]) == 0 {
+		wHeader[contentLengthHeader] = []string{strconv.FormatInt(info.Size(), 10)}
+	}
+
+	return file, info, nil
+}
+
+// serveBestFile is an upgraded version of `http.ServeFile` that primarily serves
+// a pre-compressed version of a file or its uncompressed version.
```

Comment:
If a client (say `wget`) doesn't advertise `gzip` support but tries to download an asset that is only available in pre-compressed form, will the server decompress the file on demand and serve it to the limited client? From my quick reading of the code fallback isn't implemented but I may have missed it.

Reply:
This is indeed no implemented.  As of now, the code will stop there:
```go
func (f *fileHandler) findBestFile(w http.ResponseWriter, r *http.Request, name string) (http.File, os.FileInfo, error) {
   ae := r.Header.Get(acceptEncodingHeader)
      if ae == "" { // <--- a `wget` request without explicit encoding goes inside this case. 
         // Try to read the file without any encoding.
	 return f.openAndStat(name)
   }
   
   ...
}   
```

I think that `return f.openAndStat(name)` will give an error (as all the served files are gzipped, so if the file in the query is no the `.gz` one, this will fail), therefore ending in `http.NotFound()` error

Reply:
OK then maybe pre-compression isn't something we should enable for now? Do you know how much size inflation we take by skipping the `gzip` pre-compression?

Reply:
Without pre-compression, the additional size would be around **15MB**, instead of around **4.9MB**

Reply:
And as we actually won't embed the website in the binary but rather in thee snap, this difference will not matter much right?

Reply:
Considering the snap is `xz` compressed, I suspect those 15MB will compress even better than with `gzip`. Feels like a win-win because you then don't need to care about clients not supporting `gzip` but still have a moderate snap size inflation.